promql/engine.go

```diff
@@ -3224,6 +3240,44 @@ seriesLoop:
 		}
 		s = Sample{Metric: inputMatrix[si].Metric, F: f, H: h, DropName: inputMatrix[si].DropName}
 
+		var k int64
+		var r float64
+		switch op {
+		case parser.TOPK, parser.BOTTOMK, parser.LIMITK:
+			if !convertibleToInt64(fParam) {
+				ev.errorf("Scalar value %v overflows int64", fParam)
+			}
+			k = int64(fParam)
+			if k > int64(len(inputMatrix)) {
+				k = int64(len(inputMatrix))
+			}
+			if k < 1 {
+				if ev.startTimestamp != ev.endTimestamp {
+					advanceRemainingSeries(enh.Ts, si+1)
+				}
+				return nil, annos
+			}
+		case parser.LIMIT_RATIO:
+			if math.IsNaN(fParam) {
+				ev.errorf("Ratio value %v is NaN", fParam)
+			}
+			switch {
+			case fParam == 0:
+				if ev.startTimestamp != ev.endTimestamp {
+					advanceRemainingSeries(enh.Ts, si+1)
+				}
+				return nil, annos
+			case fParam < -1.0:
+				r = -1.0
+				annos.Add(annotations.NewInvalidRatioWarning(fParam, r, e.Param.PositionRange()))
```

Comment:
Like said before for the quantile annotation, maybe this can be done once before the iteration by checking the min of all params?